tools/capture-test.py

```diff
@@ -0,0 +1,373 @@
+# energy and high frequency.  This repeats an 8-sample square wave (6
+# kHz at 48k sample rate).  Some devices can reproduce this well with
+# as few as 8 repetitions (1.3ms), but on at least one mt8195 device
+# it's unreliably audible unless repeated 128 times!  It's not caused
+# by software in the DSP, more like a codec/amp feature (possibly
+# related to power management, if we don't play other audio
+# immediately before, it's even less reliable).
+def gen_chirp_s16le(rate, chans):
+    reps = 4
+    chirp = b''
+    for i in range(opts.chirpcyc):
+        n = chans * reps
+        vals = [-0x8000] * n + [0x7fff] * n
+        chirp += struct.pack(f"{2*n}h", *vals)
+    return (chirp, opts.chirpcyc * reps)
+
+# Converts a byte array containing capture frames (which can have
+# different sample format and channel count) to the playback format
+# (always s16_le).  Also computes an "energy" value as the sum of
+# absolute sample differences (in units of +/-1.0) over all result
+# channels.  Returns both as a tuple.
+#
+# FIXME: should consider low-passing the energy computation by
+# averaging ~N recent samples.  Otherwise high frequency noise can
+# dominate, which we don't really care about measuring (AEC can't
+# treat it, and it can plausibly create false positive chirp signals
+# if loud enough).
+def cap_to_playback(buf):
+    capfmt = ('i' if opts.capbits == 32 else 'h') * opts.capchan
+    capsz = opts.capchan * (4 if opts.capbits == 32 else 2)
+    scale = 1 / (1 << (opts.capbits - 1))
+    last_frame = None
+    delta_sum = 0
+    out_frames = []
+    for i in range(0, len(buf), capsz):
+        frame = [scale * x for x in struct.unpack(capfmt, buf[i:i+capsz])[0:opts.chan]]
+        if last_frame:
+            delta_sum += sum(abs(last_frame[x] - frame[x]) for x in range(opts.chan))
+        last_frame = frame
+        iframe = [int(min(0x7fff, max(-0x8000, (1 << 15) * e))) for e in frame]
+        out_frames.append(struct.pack(f'{opts.chan}h', *iframe))
+    return (b''.join(out_frames), delta_sum)
+
+def chirp_child(wpipe):
+    for rec in pcm_do_capture(opts.duration):
+        t = rec[0]
+        (buf, energy) = cap_to_playback(rec[1])
+        frames = len(buf) / (2 * opts.chan)
+
+        # Normalize energy as "half-swing per sample" and check vs. a
+        # threshold that will trigger if we get a 0.1 unit swing over
+        # the 8-sample chirp waveform.
+        #
+        # FIXME: would be possible to do this analysis at the
+        # individual sample layer for better time fidelity instead of
+        # in 1ms chunks.
+        energy = energy / (frames * opts.chan)
+        if energy > (0.1/8):
+            os.write(wpipe, f"{t}".encode("ascii"))
+            return
+
+def echo_child(wpipe):
+    energy = 0
+    for rec in pcm_do_capture(opts.duration):
+        energy += cap_to_playback(rec[1])[1]
+
+    # Normalize energy to "half-swing per second" here, just to make
```

Comment:
not following the definition of 'half-swing', did you mean -3dBFS?

Reply:
Samples are in the range [-1,1], so one full swing is 2.0.  A single unit in the representation is half that.  But yes, 10^(-3/10) ~= 0.5, so sure.